app/src/main/java/com/vci/vectorcamapp/core/data/upload/image/ImageUploadWorker.kt

```diff
@@ -2,85 +2,508 @@ package com.vci.vectorcamapp.core.data.upload.image
+                    attempt++
+                    try {
+                        updateProgressNotification("Preparing (Attempt $attempt)...")
+                        Log.d(
+                            "ImageUploadWorker",
+                            "Uploading specimen ${specimen.id} (Attempt $attempt/$MAX_RETRIES)"
+                        )
+
+                        when (val result = performUpload(
+                            file,
+                            contentType,
+                            specimen.id
+                        )) {
+                            is DomainResult.Success -> {
+                                successfulUploads++
+                                imageUploadSuccess = true
+                                val updatedSpecimen =
+                                    specimen.copy(imageUploadStatus = UploadStatus.COMPLETED)
+                                specimenRepository.updateSpecimen(updatedSpecimen, sessionId)
+                                Log.d(
+                                    "ImageUploadWorker",
+                                    "Success for specimen ${specimen.id} on attempt $attempt"
+                                )
+                            }
+
+                            is DomainResult.Error -> {
+                                val error = result.error
+                                Log.w(
+                                    "ImageUploadWorker",
+                                    "Failed attempt $attempt for specimen ${specimen.id} with error: $error"
+                                )
+                                if (error != NetworkError.REQUEST_TIMEOUT && error != NetworkError.NO_INTERNET && error != NetworkError.SERVER_ERROR) {
+                                    Log.e(
+                                        "ImageUploadWorker",
+                                        "Non-retryable error for specimen ${specimen.id} on attempt $attempt. This is a permanent failure for this image.",
+                                    )
+                                    permanentErrorOccurred = true
+                                }
+                            }
+                        }
+                    } catch (e: Exception) {
+                        Log.e(
+                            "ImageUploadWorker",
+                            "Exception on attempt $attempt for specimen ${specimen.id}. This is a permanent failure for this image.",
+                            e
+                        )
+                        permanentErrorOccurred = true
+                    }
+                }
+
+                if (!imageUploadSuccess) {
+                    Log.e(
+                        "ImageUploadWorker",
+                        "All attempts failed or a permanent error occurred for specimen ${specimen.id}. Moving to next image."
+                    )
+                    val updatedSpecimen = specimen.copy(imageUploadStatus = UploadStatus.FAILED)
```

Comment:
Name shadowed: updatedSpecimen. Is there a cleaner/smarter way of keeping track of the status? Or is creating a copy of the specimen and then adding it to the table the best move?